File: guides/hotwire-turbo-micronaut-chat/complete/java/src/main/java/example/micronaut/services/DefaultMessageService.java
package example.micronaut.services;

import io.micronaut.context.event.ApplicationEventPublisher;
import io.micronaut.core.annotation.NonNull;
import example.micronaut.entities.Message;
import example.micronaut.models.MessageForm;
import example.micronaut.models.RoomMessage;
import example.micronaut.repositories.MessageRepository;
import example.micronaut.repositories.RoomRepository;
import jakarta.inject.Singleton;

import jakarta.validation.Valid;
import jakarta.validation.constraints.NotNull;
import java.util.Optional;

@Singleton
public class DefaultMessageService implements MessageService {
    private final MessageRepository messageRepository;
    private final RoomRepository roomRepository;
    private final ApplicationEventPublisher<RoomMessage> eventPublisher;

    public DefaultMessageService(MessageRepository messageRepository,
                                 RoomRepository roomRepository,
                                 ApplicationEventPublisher<RoomMessage> eventPublisher) { // <1>
        this.messageRepository = messageRepository;
        this.roomRepository = roomRepository;
        this.eventPublisher = eventPublisher;
    }

    @Override
    @NonNull
    public Optional<RoomMessage> save(@NonNull @NotNull @Valid MessageForm form) {
        return roomRepository.findById(form.getRoom())
                .map(room -> {
                    Message message = messageRepository.save(new Message(form.getContent(), room));
                    RoomMessage roomMessage = new RoomMessage(message.getId(),
                            form.getRoom(),
                            form.getContent(),
                            message.getDateCreated());
                    eventPublisher.publishEvent(roomMessage); // <1>
                    return roomMessage;
                });
    }
}

File: guides/hotwire-turbo-micronaut-chat/complete/python/src/example/micronaut/services/default_message_service.py
from typing import Optional

from jakarta.inject import Singleton
from micronaut.context.event import ApplicationEventPublisher

from ..entities.message import Message
from ..models.message_form import MessageForm
from ..models.room_message import RoomMessage
from ..repositories.message_repository import MessageRepository
from ..repositories.room_repository import RoomRepository
from .message_service import MessageService


# tag::clazz[]
@Singleton  # <1>
class DefaultMessageService(MessageService):
    def __init__(
        self,
        message_repository: MessageRepository,
        room_repository: RoomRepository,
        event_publisher: ApplicationEventPublisher,
    ):
        self.message_repository = message_repository
        self.room_repository = room_repository
        self.event_publisher = event_publisher

    def save(self, form: MessageForm) -> Optional[RoomMessage]:
        room = self.room_repository.findById(form.room).orElse(None)
        if room is None:
            return None

        message = self.message_repository.save(Message(content=form.content, room=room))
        room_message = RoomMessage(
            id=message.id,
            room=form.room,
            content=form.content,
            dateCreated=message.dateCreated,
        )
        self.event_publisher.publishEvent(room_message)  # <2>
        return room_message
# end::clazz[]
